### ISSUE TITLE
Applying the `preselect-function` branch to the `no-move-default` branch intended use case

### ISSUE THREAD

**Issue by gusbrs** [minad/vertico#346]
Hi @minad ,

I know you have some strong feelings about the usefulness of the idea of "preselecting the default" for completion. We did discuss this previously and you did not spare the idea. I take it you are probably tired of this one. But I think I have a good proof of concept for it, using the `preselect-function` branch. Also, you had asked me to bring to you any other relevant use cases of the branch, and this is one other idea I had, which I think is useful, and makes meaningful use of the feature. So, please bear with me.

Perhaps you have other reasons to doubt the idea of preselecting the default, but when we did discuss it, your argument was:

> Preselecting the default is pointless.  Selectrum did this and it does not lead to a better UI since as soon as you filter, the preselection and its context will be lost.

I've taken a look at how Selectrum does it, and how the `no-move-default` branch does it, and I understand your point there. When filtering starts anything "preselected" is gone, and we may get some arbitrary and large "jumps" among the candidates. However, the feature in the `preselect-function` branch, is somewhat of a misnomer. As it is built, it is actually able to dynamically reselect a candidate as filtering occurs.

That given, what I came up with is based on the following assumption:

- The relative order between any two candidates is not affected by filtering. Therefore, whatever input is in place during filtering, we can partition the remaining candidates in two sets: those which (initially) preceded the default candidate, and those which (initially) succeeded the default candidate (including itself).

If this partition is possible, we can produce a meaningful candidate selection through filtering by selecting the first candidate of the second set still available. Which, of course, will be the default, if it is still among the candidates.

With that in hand, a meaningful mechanism for (re)selecting the candidate through filtering is available, and I think this lifts your critique that "as soon as you filter, the preselection and its context will be lost". You might have other qualms with the feature, of course. But...

Anyway, I came up with the following, which needs the `preselect-function` branch (or equivalent patch), and I provisionally patched `vertico--move-to-front` to avoid moving the default candidate, just because this seemed the easiest for a proof of concept such as this one. Also, I used `TeX-insert-macro` and `TeX-command-master` because these were two commands where I particularly missed such behavior. I'm not even sure `vertico-multiform` is the best way to go for such a feature. Also, very lightly tested so far, though if felt right for as much as I did test. But, again, this is just proof of concept for the basic mechanism.

```elisp
(with-eval-after-load 'vertico
  ;; Experimenting with 'no-move-default' branch and default preselection
  ;; based on the 'preselect-function' branch.

  ;; Make Vertico "not move the default", that's what the branch does, and
  ;; the right place to intervene for this purpose is probably
  ;; 'vertico--recompute' but, provisionally, since I don't want to mix
  ;; things with the preselect-function stuff above, I'll do it in
  ;; 'vertico--move-to-front'.
  (defcustom gb/vertico-preselect-default nil
    "")
  (el-patch-defun vertico--move-to-front (elem list)
    "Move ELEM to front of LIST."
    (if-let (found
             (el-patch-swap
               (member elem list)
               (and (not gb/vertico-preselect-default) (member elem list))))
        (nconc (list (car found)) (delq (setcar found nil) list))
      list))

  (defvar-local gb/vertico-preselect-post-default nil)
  (defvar-local gb/vertico-preselect-post-default-stored nil)
  (defun gb/vertico-preselect-default (cands)
    (unless gb/vertico-preselect-post-default-stored
      (let ((def (or (car-safe minibuffer-default) minibuffer-default)))
        (setq gb/vertico-preselect-post-default (or (member def cands) cands)))
      (setq gb/vertico-preselect-post-default-stored t))
    (or (when (= 0 (length cands)) -1)
        (let ((index 0))
          (catch 'cand-found
            (dolist (e cands)
              (when (member e gb/vertico-preselect-post-default)
                (throw 'cand-found index))
              (setq index (1+ index)))
            nil))
        (1- (length cands))))

  (add-to-list 'vertico-multiform-commands
               '(TeX-insert-macro
                 (vertico-sort-function . identity)
                 (gb/vertico-preselect-default . t)
                 (vertico-preselect . gb/vertico-preselect-default)))

  (add-to-list 'vertico-multiform-commands
               '(TeX-command-master
                 (vertico-sort-function . identity)
                 (gb/vertico-preselect-default . t)
                 (vertico-preselect . gb/vertico-preselect-default))))
```

WDYT?

**Comment by minad** [comment]
> However, the feature in the preselect-function branch, is somewhat of a misnomer.

The name may be a bit confusing. By the current definition, preselect happens after filtering *before* moving up or down.

> The relative order between any two candidates is not affected by filtering. Therefore, whatever input is in place during filtering, we can partition the remaining candidates in two sets: those which (initially) preceded the default candidate, and those which (initially) succeeded the default candidate (including itself).

Sure, I wrote about that in some other issue. However I don't get your argument or I think you don't bring anything new to the table. The order of candidates is usually not meaningful in completion. There are only a handful of special commands which are an exception, e.g., consult-line. For these commands the candidates are derived from the buffer or have otherwise some natural ordering *and* the default candidate is naturally not the topmost candidate.

> I take it you are probably tired of this one. 

I still have this proposal in mind, however I don't have the bandwidth right now to think about it again. We can revisit this in a while.

**Comment by gusbrs** [comment]
Thanks for considering. Cheers!

**Comment by minad** [comment]
> Also, I used TeX-insert-macro and TeX-command-master because these were two commands where I particularly missed such behavior. 

In the meantime, could you please elaborate on why you miss the particular behavior for these two commands? I didn't understand that above. Do these commands satisfy the criteria (natural ordering and default not at the top).

**Comment by gusbrs** [comment]
> In the meantime, could you please elaborate on why you miss the particular behavior for these two commands? I didn't understand that above. Do these commands satisfy the criteria (natural ordering and default not at the top).

Well, they do have some sorting which, once you get used to it, is a source of information. `TeX-insert-macro` is sorted at the source, using `TeX-car-string-lessp`, which uses `string-lessp` so it's more or less "alphabetical" order (and I personally change this to use `string-collate-lessp` so to get case insensitive sorting). One "personal idiom" which I miss when the default is moved to the top is for inserting some common pairs, for example, `\makeatletter` / `\makeatother` or `\ExplSyntaxOn` / `\ExplSyntaxOff`. If the default is moved to the top I have to type in for filtering for both of them. Whereas, if the order is preserved and the default selected, I only need to filter once, so the usual in this case would be: `C-c RET makeat RET RET C-c RET C-n RET`. Done. Sometimes, inserting citation commands, they are close to each other, and so on.

In the case of `TeX-command-master` this is a very short list (say, a dozen candidates) which I personally manually sort the 3-4 candidates which I use the most, placing them on the top, and usually just move point to select. So, in this case, it is actually jarring when the default is moved to the top, because I expect a certain order. This same sort of expectation may also be relevant for the muscle memory of `vertico-indexed`, for example.

You may say these are atypical, and you might be right, but there you have why I'm trying to get this working.  ;-)

**Comment by minad** [comment]
> Whereas, if the order is preserved and the default selected, I only need to filter once, so the usual in this case would be: C-c RET makeat RET RET C-c RET C-n RET. Done. Sometimes, inserting citation commands, they are close to each other, and so on.

Just for me to understand this correctly - the default candidate for the second invocation is the one you selected last? This is an argument, however I wouldn't call it "natural" as in the case of `consult-line`. It seems rather like a happy accident or not so happy accident without `vertico-preselect`.
 
> `TeX-command-master` ... `vertico-indexed`

In this case, I think something is wrong with `TeX-command-master`. The default command should be at the top in the first place, which would make the movement a non-issue. I recommend to rely on `vertico-indexed` for muscle memory. Completion is the wrong tool then. One should better use a prefix map instead.

**Comment by gusbrs** [comment]
> Just for me to understand this correctly - the default candidate for the second invocation is the one you selected last? This is an argument, however I wouldn't call it "natural" as in the case of `consult-line`. It seems rather like a happy accident or not so happy accident without `vertico-preselect`.

Yes, this presumes that, in the second invocation the default hasn't been moved, but rather preselected, as the snippet above provides for. Whether it is "natural", I don't know, it is just useful to leverage an order of candidates which is known and can be relied on. And it's not an "accident", it is a given order which is there provided no one changes it.

> In this case, I think something is wrong with TeX-command-master. The default command should be at the top in the first place, which would make the movement a non-issue.

This is just the way I use it. As I said, I manually sort its elements. But, by default, I don't think it, by itself, sorts the default on top, it is just the default. The order is just the order in which the user option / variable is set. And, of course, I only move when I want something different from the default, otherwise, just RET.

> I recommend to not rely on vertico-indexed for muscle memory. Completion is the wrong tool then. One should better use a prefix map instead.

You're probably right there. But the sorting information is there, and could be relied upon, if Vertico did not tamper with it. Of course, I only mean here the moving of the default candidate when the sort function is `identity`. When it's not, the sorting is obviously as desired.

**Comment by minad** [comment]
> And it's not an "accident", it is a given order which is there provided no one changes it.

My point is that you are relying on specific namings where related commands are accidentally next to each other.

> But the sorting information is there, and could be relied upon, if Vertico did not tamper with it.

The equally simple alternative would be to use the Selectrum behavior which I find worse due to the jumping during filtering.

One could make the movement of the default actually part of the sort function such that it wouldn't occur if sorting is overridden. Then one would also need the Selectrum-like default selection on top. Maybe that would be better than the current solution but maybe it is also too late to change this now.

But again, I think you are doing something wrong if you rely on candidate order and don't even use filtering. You simply use completion as a menu then, where prefix maps or other means are better tools. Sorry for repeating myself.

**Comment by gusbrs** [comment]
> My point is that you are relying on specific namings where related commands are accidentally next to each other.

My only point is that these cases do happen and the order of the candidates carries information. Also that moving the default to the top of the list wastes such information.

> The equally simple alternative would be to use the Selectrum behavior which I find worse due to the jumping during filtering.

Yes, pretty much that's something of the sort which I'm trying to achieve. And what I thought I was bringing to the table with the above is something similar to that, but without "jumping during filtering" (or at least less of it).

> One could make the movement of the default actually part of the sort function such that it wouldn't occur if sorting is overridden. Then one would also need the Selectrum-like default selection on top. Maybe that would be better than the current solution but maybe it is also too late to change this now.

That'd be interesting I think. But regarding "too late", again, I'm just suggesting for some handles to be available for interested users to change the default behavior. I had made this point regarding the `preselect-function` and I guess I'm adding to the list some form to override the hard-coded movement of the default candidate. Whether this is done through a variable or by integrating it to the sort function, I'd just be happy if it exists.

> But again, I think you are doing something wrong if you rely on candidate order and don't even use filtering. You simply use completion as a menu then, where prefix maps or other means are better tools. Sorry for repeating myself.

I'm sorry to make you repeat yourself. ;-) It is a fact that I'm way behind in these waters, that's why I asked for your patience in opening the issue. So perhaps I'm wrong in relying in candidate order. But, perhaps for force of habit, I do tend to think this is an information I can use, but Vertico does break this expectation. That's why I'm trying to bend it for my preferences. That's as much as I can claim.

**Comment by gusbrs** [comment]
Not insisting, but only for for the record, since you are interested in the use cases. I just remembered one other place where I expect a certain order, and thus dislike the default to be moved: `org-refile` (and friends).

**Comment by gusbrs** [comment]
Just reporting back on this one. After experimenting a while with this, I think this preselection function (`gb/vertico-preselect-default`, that is) is not such a good one. The reason for this is that, by partitioning the candidates between before and after the original place where the default candidate is, the preselection becomes dependent on the position of the default candidate, even when it is no longer among the candidates. So, the selected candidate for a given input is not deterministic, depending on the position relative to the default candidate.

That given, I moved to a simpler preselection function which just selects the default candidate if it is present, otherwise, from top. Which (I think) is what `selectrum` does.

```elisp
(defun gb/vertico-preselect-default-selectrum (cands)
  (let ((def (or (car-safe minibuffer-default) minibuffer-default)))
    (or (when (= 0 (length cands)) -1)
        (and def (seq-position cands def))
        0)))
```
Jumpier, yes, but I do like the result.

Don't get me wrong, I still think the ability to not move the default candidate is useful, with a proper preselection function. I'm just acknowledging this particular strategy of preselection I came up with was misguided.

**Comment by minad** [comment]
I think you can always use the value of `minibuffer-default` (if non-nil and if it is a string), and then search the candidate just next to the default. You can use the `consult-location` text property for that. Did I misunderstand your original strategy?

> The reason for this is that, by partitioning the candidates between before and after the original place where the default candidate is, the preselection becomes dependent on the position of the default candidate, even when it is no longer among the candidates. So, the selected candidate for a given input is not deterministic, depending on the position relative to the default candidate.

Yes, but wasn't this the goal all along? You wanted to find the next candidate starting from the default (= the current line). This is how Swiper supposedly works.

Anyway, this obviously needs more thought. There is a reason why I am not moving fast on this one.

**Comment by gusbrs** [comment]
> I think you can always use the value of `minibuffer-default` (if non-nil and if it is a string), and then search the candidate just next to the default. You can use the `consult-location` text property for that. Did I misunderstand your original strategy?

The function was doing what was intended, identifying the candidate "next to the default" was working all right, but that's not the problem.

> Yes, but wasn't this the goal all along? You wanted to find the next candidate starting from the default (= the current line).

It was, I thought it was a good idea, and I'm granting it wasn't.

> This is how Swiper supposedly works.

Well, I think so, and this is indeed similar to what I got for `consult-line` with https://github.com/minad/consult/issues/287#issuecomment-1466158081. But I don't think that's how Ivy (as opposed to Swiper) in general works (having used it for years, I sincerely don't remember, I guess I'm getting more aware of this of late... But I just did a quick test here, and I think what it does is similar to Selectrum, and to this other function I just posted, in the absence of the default, filtering continues from the top).

> Anyway, this obviously needs more thought. There is a reason why I am not moving fast on this one.

I know, I was just acknowledging a bad idea because it was due.  ;-)

**Comment by gusbrs** [comment]
@minad In thinking about this one today, and testing a little with Selectrum and Ivy, I understood better the multiple uses of `vertico--move-to-front` inside `vertico--recompute` and I was able to generalize what I had, to what I think are "shareable results", a little beyond the "proof of concept" initially provided.

Now, admittedly this does not add anything new to the discussion regarding the moving of the default candidate to the top. It is what you already know. Best result, as far as I know, is Selectrum, and I know what you think of its limitations in this regard. To the best of my knowledge, the following preselection function does what Selectrum with `selectrum-move-default-candidate` set to `nil` does with regard to the default candidate (keeps it preselected during filtering, as long as it is among the candidates, but when it's not, filter from the top), and does what Ivy does with regard to exact matches (preselects it, instead of moving it to the top, as Vertico and Selectrum do).

I'm only bringing this because one of the intentions on opening this issue was providing a further case of meaningful use of the `preselect-function` branch -- something I had promised you to do -- and, properly generalized, I think this function does (despite the OP not having been able to improve on the approach to the default candidate).

```elisp
(defvar-local gb/vertico-no-move-to-front nil)
(defun gb/vertico--move-to-front-advice (orig-fun elem list)
  (if gb/vertico-no-move-to-front
      list
    (funcall orig-fun elem list)))
(advice-add 'vertico--move-to-front
            :around #'gb/vertico--move-to-front-advice)

(defun gb/vertico-preselect-no-move-to-front (all)
  (let* (;; Reconstructing the args and let bound variables from
         ;; `vertico--recompute', see also `vertico--update'.
         (pt (max 0 (- (point) (minibuffer-prompt-end))))
         (content (minibuffer-contents-no-properties))
         (table minibuffer-completion-table)
         (pred minibuffer-completion-predicate)
         (before (substring content 0 pt))
         (after (substring content pt))
         (bounds (or (condition-case nil
                         (completion-boundaries before table pred after)
                       (t (cons 0 (length after))))))
         (field (substring content (car bounds) (+ pt (cdr bounds))))
         (completing-file (eq 'file (vertico--metadata-get 'category)))
         (def (or (car-safe minibuffer-default) minibuffer-default))
         (def-missing (and def (equal content "") (not (member def all)))))
    (cond ((or def-missing (not all))
           -1)
          ((not gb/vertico-no-move-to-front)
           0)
          ((seq-position all field))
          ((and completing-file (not (string-suffix-p "/" field))
                (seq-position all (concat field "/"))))
          ((and (stringp def)
                (seq-position all def)))
          (t 0))))

(dolist (cmd '(org-refile
               org-agenda-refile
               org-capture-refile
               TeX-insert-macro
               TeX-command-master))
  (add-to-list 'vertico-multiform-commands
               `(,cmd
                 (vertico-sort-function . identity)
                 (gb/vertico-no-move-to-front . t)
                 (vertico-preselect . gb/vertico-preselect-no-move-to-front))))
``` 
This generalizes the previous `gb/vertico-preselect-default-selectrum` in handling all uses of `vertico--move-to-front`, such that that advice to it is more than a provisional hack just to play with a proof of concept attempt. It also takes care to reproduce the behavior of `vertico--recompute` in all cases which are not those affected by the preselection of either the default or of an exact match. Still lightly tested though.

One side benefit of this, if and when you decide to merge the `preselect-function` branch, is that I could make this available in the wiki, thus hopefully appeasing those interested in such behavior, and in the meantime you can get some peace in making your call about the built-in behavior of the default candidate. ;-)

**Comment by minad** [comment]
As I see it, there are two acceptable choices, either moving the default and exact match to the top or preselect them. I made the first choice for Vertico. It is better since you see the selected candidate directly next to the prompt, which makes it easier to focus. Furthermore it is advantageous since we avoid jumping and we get simpler preselection logic. The choice is worse since it destroys the order. There are more arguments in favor than against. My suggestion is that you try to use my packages as I designed them. They work fine as is. Also, my packages are somewhat opinionated and I intend to keep their style. :)
